File: 2024-11-November-GFG-POTD/November 10 - Union of Two Sorted Arrays with Distinct Elements/Solution.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.lang.*;
import java.util.*;

class Main {
    public static void main(String args[]) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int t = Integer.parseInt(br.readLine().trim());
        while (t-- > 0) {
            // First array input
            String[] str1 = br.readLine().trim().split(" ");
            int n = str1.length;
            int[] a = new int[n];
            for (int i = 0; i < n; i++) {
                a[i] = Integer.parseInt(str1[i]);
            }

            // Second array input
            String[] str2 = br.readLine().trim().split(" ");
            int m = str2.length;
            int[] b = new int[m];
            for (int i = 0; i < m; i++) {
                b[i] = Integer.parseInt(str2[i]);
            }

            Solution obj = new Solution();
            ArrayList<Integer> res = new ArrayList<Integer>();
            res = obj.findUnion(a, b);
            for (int i = 0; i < res.size(); i++) System.out.print(res.get(i) + " ");
            System.out.println();
        }
    }
}

// } Driver Code Ends

// User function Template for Java

class Solution {
    // Function to return a list containing the union of the two arrays.
    public static ArrayList<Integer> findUnion(int a[], int b[]) {
        ArrayList<Integer> result = new ArrayList<>();
        int i = 0, j = 0;
        
        // Traverse both arrays
        while (i < a.length && j < b.length) {
            // Avoid duplicates in result
            if (!result.isEmpty() && result.get(result.size() - 1) == a[i]) {
                i++;
                continue;
            }
            if (!result.isEmpty() && result.get(result.size() - 1) == b[j]) {
                j++;
                continue;
            }
            
            if (a[i] < b[j]) {
                result.add(a[i]);
                i++;
            } else if (a[i] > b[j]) {
                result.add(b[j]);
                j++;
            } else {
                result.add(a[i]);
                i++;
                j++;
            }
        }
        
        // Add remaining elements of a
        while (i < a.length) {
            if (result.isEmpty() || result.get(result.size() - 1) != a[i]) {
                result.add(a[i]);
            }
            i++;
        }
        
        // Add remaining elements of b
        while (j < b.length) {
            if (result.isEmpty() || result.get(result.size() - 1) != b[j]) {
                result.add(b[j]);
            }
            j++;
        }
        
        return result;
    }
}

File: 2024-11-November-GFG-POTD/November 10 - Union of Two Sorted Arrays with Distinct Elements/Solution.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    // Function to return a vector containing the union of the two arrays.
    vector<int> findUnion(int a[], int b[], int n, int m) {
        vector<int> result;
        int i = 0, j = 0;

        // Traverse both arrays
        while (i < n && j < m) {
            // Avoid duplicates in result
            if (!result.empty() && result.back() == a[i]) {
                i++;
                continue;
            }
            if (!result.empty() && result.back() == b[j]) {
                j++;
                continue;
            }

            if (a[i] < b[j]) {
                result.push_back(a[i]);
                i++;
            } else if (a[i] > b[j]) {
                result.push_back(b[j]);
                j++;
            } else {
                result.push_back(a[i]);
                i++;
                j++;
            }
        }

        // Add remaining elements of a
        while (i < n) {
            if (result.empty() || result.back() != a[i]) {
                result.push_back(a[i]);
            }
            i++;
        }

        // Add remaining elements of b
        while (j < m) {
            if (result.empty() || result.back() != b[j]) {
                result.push_back(b[j]);
            }
            j++;
        }

        return result;
    }
};

int main() {
    int t;
    cin >> t;
    while (t--) {
        // First array input
        int n;
        cin >> n;
        int a[n];
        for (int i = 0; i < n; i++) {
            cin >> a[i];
        }

        // Second array input
        int m;
        cin >> m;
        int b[m];
        for (int i = 0; i < m; i++) {
            cin >> b[i];
        }

        Solution obj;
        vector<int> res = obj.findUnion(a, b, n, m);
        for (int i = 0; i < res.size(); i++) cout << res[i] << " ";
        cout << endl;
    }
    return 0;
}
